src/memory.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2019 IBM Corp. and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "winter_config.h"
+
+#ifdef WINTER_USE_MMAP
+#include <sys/mman.h>
+#endif
+
+#include "memory.hpp"
+
+namespace winter {
+
+Memory::~Memory() {
+#ifdef WINTER_USE_MMAP
+    if (_data.start)
+        munmap(_data.start, _data.current_capacity_pages.get() << WASM_PAGE_SHIFT);
+#else
+    std::free(_data.start);
+#endif
+}
+
+bool Memory::alloc_exactly(NumPages num_pages) {
+    WASSERT((_data.flags & MEMORY_FLAG_SHARED) == 0, "Shared WebAssembly memory cannot be grown");
```

Comment:
Can this be turned into a call to `is_shared()`?

```suggestion
    WASSERT(is_shared(), "Shared WebAssembly memory cannot be grown");
```

Reply:
Yeah, I've gone ahead and replaced that. For context, this method used to be on `MemoryInternal` in a previous iteration of this PR, which didn't have `is_shared()`.